### PR TITLE
feat: add HDD monitoring items to hdd_monitor

### PR DESCRIPTION
**Pull request by v-nakayama7440-esol**
Signed-off-by: v-nakayama7440-esol <v-nakayama7440@esol.co.jp>

## Description

Add the following information to be monitored in `hdd_monitor` based on safety analysis made by TIER IV.
When a value exceeds threshold, the `hdd_monitor` generates `Warning` to notify operator something is happening on a storage.

- **HDD RecoveredError**
  - Observe a CRC error on the storage to monitor Soft error(Transient fault).
- **HDD ReadDataRate**
  - Observe Read data rate to monitor High load on communication bandwidth.
- **HDD WriteDataRate**
  - Observe Write data rate to monitor High load on communication bandwidth.
- **HDD ReadIOPS**
  - Observe input/output per second for read to monitor High load on communication bandwidth.
- **HDD WriteIOPS**
  - Observe input/output per second for write to monitor High load on communication bandwidth.

**_(Generally, it would be better to use official value for threshold in datasheet if provided from HDD vendor. If a specification is not provided from HDD vendor, it would be necessary to use a peak value as results of benchmark testing.)_**

In addition, we will improve `hdd_monitor` functionality as regards to HDD connection.

- **HDD Connection**
  - Newly added diagnostic information on HDD connection. 
  - The `hdd_monitor` generates `Warning` if a device is not connected because of the availability of Autoware.
  - Generally, a built-in SSD is not disconnected from a computer, it is not a problem. However, an external SSD may be disconnected while running Autoware(driving/stopping) and the current implementation of `hdd_monitor` generates `Error` on every diagnostic information if a device is not connected. This leads to an emergency stop and loss of availability of Autoware.
  Maybe we should restart Autoware just for disconnecting an external SSD while running Autoware.
  - There is an issue where the filesystem remains mounted when the drive is actually disconnected. When HDD Connection detects this issue, hdd_reader will do a delayed unmount.

- Do not show any information on each diagnostic information when a device is not connected.
  - This will allow us to understand a error is happening due to limit exceeded.
  - The current implementation generates  `Error` when a device is not connected in addition to limit exceeded, so the cause of a error is mixed.

## Related links
https://github.com/autowarefoundation/autoware.universe/issues/1770

## Tests performed
We try to use peak values as results of benchmark testing for this test.

### Preparation
1. Install KDiskMark
   ```console
   sudo add-apt-repository ppa:jonmagon/kdiskmark
   sudo apt install kdiskmark
   ```

1. Run KDiskMark.
   ```console
   kdiskmark
   ```

1. Click `All` to start benchmark testing.
Screenshot from 2022-09-06 05-34-30

1. You can see the following results.
   - `MB/s` for ReadDataRate and WriteDataRate
Screenshot from 2022-09-05 17-41-58
   - `IOPS` for ReadIOPS and WriteIOPS
Screenshot from 2022-09-06 05-33-54

1. Change thresholds in `hdd_monitor.param.yaml` as follows.
    Decimals shall be rounded down.
```yaml
        ead_data_rate_warn: 502.0 # MB/s
        write_data_rate_warn: 109.0 # MB/s
        read_iops_warn: 44209.0 # IOPS
        write_iops_warn: 16970.0 # IOPS
```

### Verify every information report `OK` in normal operation.
1. Run Autoware.

1. Run rqt_runtime_monitor. 
    ```console
    ros2 run rqt_runtime_monitor rqt_runtime_monitor
    ```

1. Verify net_monitor: every information report `OK`.
   1. **HDD RecoveredError**
image

   1. **HDD ReadDataRate**
image

   1. **HDD WriteDataRate**
image

   1. **HDD ReadIOPS** 
image

   1. **HDD WriteIOPS**
image

### Verify no warning and error messages in a normal run.
image

### Verify every information are aggregated in `resource_monitoring`.
1. Run rqt_robot_monitor.
   ```console
   ros2 run rqt_robot_monitor rqt_robot_monitor
   ```

1. Verify the following items are in `storage`
    - recovered_error
    - read_data_rate
    - write_data_rate
    - read_iops
    - write_iops
    - connection
image

### Verify each information reports `Warning` in abnormal operation.

1. Run KDiskMark.
   ```console
   kdiskmark
   ```

1. Verify each information regarding data rate reports `Warning`.

   1. Click `SEQ1M Q8T1` on KDiskMark window.
image

   1. **HDD ReadDataRate**
image

   1. **HDD WriteDataRate**
image

1. Verify each information regarding IOPS reports `Warning`.

   1. Click `RND4K Q32T1` on KDiskMark window.
image

   1. **HDD ReadIOPS**
image

   1. **HDD WriteIOPS**
image

    #### _**HDD RecoveredError**_
    - ECC error may not be supported by some devices. In that case, this will be reported as `OK` with `not available` message.
    - ECC error is generally caused by storage degradation, and there is no way to cause ECC error intentionally, so the test can be skipped as regards to HDD RecoveredError.

### Verify `hdd_monitor: HDD Connection` reports `OK` when an external storage is connected.

1. Connect an external storage to the computer.

1. Verify `hdd_monitor: HDD Connection` reports `OK`.
image

1. Verify `hdd_monitor` is showing information about external storage on each diagnostic information.
    ex.)
image

### Verify `hdd_monitor: HDD Connection` reports `OK` when an external storage is disconnected.

1. Disconnect an external storage from the computer.

1. Verify `hdd_monitor: HDD Connection` reports `Warning`.
**NG. Reporting Error**
image

1. Verify `hdd_monitor` does not show any information about external storage (which is disconnected) on each diagnostic information.
    ex.)
image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by ito-san** [comment]
@v-nakayama7440-esol Please fix the error in pre-commit.ci. and I intend to file a new issue to explain why multiple PRs related to diagnosis are submitted.

**Comment by v-nakayama7440-esol** [comment]
@ito-san Fixed pre-commit.ci error.
fb6ca488bc89749adbfe494235f4e5446db93483

**Comment by v-nakayama7440-esol** [comment]
@ito-san @yukkysaito 
Fixed the monitoring method of RecoveredError. 

I used the value of RecoveredError obtained from the device as it is. However, it is correct to use the increment of RecoveredError since launching the app.
a312a40a0eae84aa6adc783a2730c7505ba78457

**Comment by ito-san** [comment]
@v-nakayama7440-esol 
- `hdd_monitor: HDD Connection` is still reporting `OK` after disconnecting an USB SSD.
image
- `Error` is reported on each information of `hdd_monitor`.
image
Could you investigate and fix this?

File: system/system_monitor/src/hdd_monitor/hdd_monitor.cpp
Comment (ito-san) [inline system/system_monitor/src/hdd_monitor/hdd_monitor.cpp:829]: @v-nakayama7440-esol 
This will be failed with unprivileged user. Please think of another way.

**Review by ito-san — CHANGES_REQUESTED** [review]

File: system/system_monitor/src/hdd_monitor/hdd_monitor.cpp
Comment (ito-san) [inline system/system_monitor/src/hdd_monitor/hdd_monitor.cpp:818]: @v-nakayama7440-esol 
I'm sorry but it's difficult to understand this.
The commend should be as follows:
`Deal with the issue that file system remains mounted when a drive is actually disconnected.`

**Review by ito-san — CHANGES_REQUESTED** [review]

File: system/system_monitor/src/hdd_monitor/hdd_monitor.cpp
Comment (v-nakayama7440-esol) [reply system/system_monitor/src/hdd_monitor/hdd_monitor.cpp]: @ito-san 
I moved unmount function to hdd_reader.
b679b8a57aff29f39d265a14a8892382e4d0bc14

File: system/system_monitor/src/hdd_monitor/hdd_monitor.cpp
Comment (v-nakayama7440-esol) [reply system/system_monitor/src/hdd_monitor/hdd_monitor.cpp]: @ito-san 
I fixed the comment.
b679b8a57aff29f39d265a14a8892382e4d0bc14

File: system/system_monitor/include/hdd_reader/hdd_reader.hpp
Comment (ito-san) [inline system/system_monitor/include/hdd_reader/hdd_reader.hpp:105]: @v-nakayama7440-esol
Sharing request for S.M.A.R.T. information and request to lazy unmount in one struct would not be good practice because there would be a risk for this struct to be bigger and bigger for further improvement and this struct would be hard to be managed. Maybe significant changes, but these functionalities should be separated into message ids based on usage.

**Review by ito-san — CHANGES_REQUESTED** [review]

File: system/system_monitor/include/hdd_reader/hdd_reader.hpp
Comment (v-nakayama7440-esol) [reply system/system_monitor/include/hdd_reader/hdd_reader.hpp:108]: @ito-san 
I separated requests for S.M.A.R.T. and lazy unmounts.
c97b8bc4436976d2d76d0efc701bf5ba54404eec

File: system/system_monitor/src/hdd_monitor/hdd_monitor.cpp
Comment (ito-san) [reply system/system_monitor/src/hdd_monitor/hdd_monitor.cpp]: Verified the issue was gone after the fix.

File: system/system_monitor/include/hdd_reader/hdd_reader.hpp
Comment (ito-san) [reply system/system_monitor/include/hdd_reader/hdd_reader.hpp:108]: Thanks.

**Comment by ito-san** [comment]
@v-nakayama7440-esol Would you resolve conflicts in `system/system_monitor/README.md`?

**Comment by v-nakayama7440-esol** [comment]
@ito-san 
I resolved conflict.

**Review by ito-san — APPROVED** [review]

**Comment by ito-san** [comment]
@kenji-miyake Would you please review this PR because you are the code owner of `tier4_system_launch` and  `system_error_monitor`. Looks like your approval is required before merging.

**Comment by kenji-miyake** [comment]
@ito-san Could you add Ito-san and other necessary members here?
https://github.com/autowarefoundation/autoware.universe/blob/369b75192c89828d3582a7aa96f44b80c8272216/launch/tier4_system_launch/package.xml#L7

**Comment by ito-san** [comment]
> @ito-san Could you add Ito-san and other necessary members here?
> 
> https://github.com/autowarefoundation/autoware.universe/blob/369b75192c89828d3582a7aa96f44b80c8272216/launch/tier4_system_launch/package.xml#L7

OK, I will try to file a PR.

**Review by kenji-miyake — APPROVED** [review]
I've reviewed `system.param.yaml`, LGTM.

**Pull request merged**
